### ISSUE TITLE
SMAPI raises graphics events after sprite batch's render target has been removed (for some users)

### ISSUE THREAD

**Issue by Pathoschild** [ClxS/SMAPI#151]
For a subset of users, SMAPI raises graphics events after drawing to the game's sprite batch will no longer have any effect.
## Background

SMAPI draws to the screen in one of two ways:
- At 100% zoom, it draws directly to the output. This seems to work fine for all users.
- At any other zoom, it...
  1. redirects rendering to a temporary render target (e.g. [lines 855–858](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L855-L858));
  2. draws everything to that target (e.g. [line 1306](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L1306)) and ends the sprite batch (e.g. [line 1316](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L1316));
  3. draws that target to the screen and discards it (e.g. [lines 1320–1327](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L1320-L1327)).

Importantly, it raises `OnPostRenderGuiEvent` (e.g. [line 1307](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L1307)) and `OnPostRenderEvent` (e.g. [line 1315](https://github.com/cjsu/SMAPI/blob/b7e58da/src/StardewModdingAPI/Inheritance/SGame.cs#L1315)) _before_ ending the sprite batch and discarding the temporary render target.
## The issue

For a subset of users:
- Anything drawn in the `OnPostRenderGuiEvent`, `OnPostRenderHudEvent`, and `OnPostRenderEvent` events at a non-100% zoom is invisible. At 100% zoom, it's shown correctly.
- This happens consistently — some users _never_ see the drawn UI at a non-100% zoom; others _always_ see it. There are no reports of inconsistent behaviour.

Notes based on testing:
- For affected users at a non-100% zoom when drawing within graphics event handlers, `Game1.spriteBatch.GraphicsDevice.GetRenderTargets()` is consistently empty. For unaffected users, it consistently has one value (presumably `SGame::Screen`).
- Confirmed with the latest version of SDV (1.1.1) and SMAPI (0.40 1.11-3) on Windows. May also occur with SMAPI 0.40 1.11 (based on earlier reports for which I no longer have logs).
- Confirmed that it affects multiple mods ([Chests Anywhere 1.6](https://github.com/Pathoschild/ChestsAnywhere/tree/1.6) and [1.7](https://github.com/Pathoschild/ChestsAnywhere/tree/49aab1f), and [Loved Labels 2.0](http://www.nexusmods.com/stardewvalley/mods/279/)).
- Confirmed that it happens even if only Chests Anywhere in installed, so there's no mod conflict.
## Reproducing the issue

I haven't found any way of deliberately causing the issue — it consistently works for some users, consistently breaks for others. However, it's possible to test it.
1. Install the latest version of SMAPI.
2. Install [this custom mod](https://github.com/cjsu/SMAPI/files/546160/MinimalOverlay-1.0.zip) created specifically to test this issue (see 
   [source code](https://github.com/cjsu/SMAPI/files/546161/MinimalOverlay-1.0-source-code.zip)).
3. Load a game with zoom level=100%.
4. You should see a red overlay with crosshairs:  
   ![overlay](https://cloud.githubusercontent.com/assets/230581/19624039/0aa45eaa-98b3-11e6-99c9-1e69ca114d38.png)
5. Zoom out.
   - If you're unaffected, you should still see the overlay.
   - If you're affected, the overlay will disappear.

The mod will log relevant metadata to the console.
- For an unaffected user, note that a render target (presumably `SGame::Screen`) is added when the player zoomes out:
  
  ```
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=1 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.95 and 1 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.9 and 1 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.85 and 1 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.8 and 1 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.75 and 1 renderTargets
  ```
- For an affected user, note that the render targets are empty at any zoom level:
  
  ```
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=1 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.95 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.9 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.85 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.8 and 0 renderTargets
  [MinimalOverlay] Handling OnPostRenderEvent with zoomLevel=0.75 and 0 renderTargets
  ```
## Known affected users
- coffeefox on the [Stardew Valley Discord](https://discord.gg/KCJHWhX) (version 0.40.1.1-3);
- starried via [Nexus Mods bug report](http://www.nexusmods.com/stardewvalley/mods/518/?tab=14&navtag=http%3A%2F%2Fwww.nexusmods.com%2Fstardewvalley%2Fajax%2Fmodbugs%2F%3Fid%3D518%26openreport%3D185100) (version 0.40.1.1);
- TheExile2 via [Nexus Mods bug report](http://www.nexusmods.com/stardewvalley/mods/518/?tab=14&navtag=http%3A%2F%2Fwww.nexusmods.com%2Fstardewvalley%2Fajax%2Fmodbugs%2F%3Fid%3D518%26openreport%3D183795) (possibly; fixed by updating from 0.40.1.1 to 0.40.1.1-3?).

(Thanks coffeefox for letting me use you as a guinea pig, since I'm not affected. 😄)

**Comment by Pathoschild** [comment]
None of the SMAPI developers can reproduce this, so we need volunteers affected by this issue. Moving this into the backlog until we can find an affected user.

**Comment by Pathoschild** [comment]
Tested with an affected user; this seems to be fixed in 1.0. 🎊
